### PR TITLE
allow users to configure basic authentication credentials

### PR DESCRIPTION
**Pull request by yanshiyason**
re #26

Sorry, I had to open a new PR because I wanted to change the branch on my fork.

### CODE REVIEW

File: Sources/TurboNavigator/TurboNavigator.swift [hunk 1]
```diff
@@ -239,4 +242,8 @@ extension TurboNavigator: SessionDelegate {
     }
 
     public func sessionWebViewProcessDidTerminate(_ session: Turbo.Session) {}
+
+    public func session(_ session: Session, didReceiveAuthenticationChallenge challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
+      delegate?.session(session, didReceiveAuthenticationChallenge: challenge, completionHandler: completionHandler)
```
Comment (joemasilotti) [inline Sources/TurboNavigator/TurboNavigator.swift]:
```suggestion
        delegate?.session(session, didReceiveAuthenticationChallenge: challenge, completionHandler: completionHandler)
```

**Review by joemasilotti — CHANGES_REQUESTED** [review]
Can you please confirm that if a basic auth dialog is presented we are raising/showing the same exception vs. if someone wired up a `WKWebView` without Turbo Navigator?

**Comment by yanshiyason** [comment]
Hi Joe, sorry for the delay, I will pick this back up as soon as possible.

**Comment by yanshiyason** [comment]
@joemasilotti Here are the results of my research. Please have a look and let me know how you think we should proceed 🙇 

# UIWebView

Here is how I tested the UIWebView case:

In the `SceneDelegate` I prepared a view controller with a web view and pointed the web view to the basic auth protected website.

```swift
let vc = UIViewController()
let view = vc.view!
let webView = WKWebView()

webView.load(URLRequest(url: URL(string: "https://jigsaw.w3.org/HTTP/Basic/")!))
view.addSubview(webView)
webView.translatesAutoresizingMaskIntoConstraints = false
webView.topAnchor.constraint(equalTo: view.topAnchor).isActive = true
webView.bottomAnchor.constraint(equalTo: view.bottomAnchor).isActive = true
webView.leadingAnchor.constraint(equalTo: view.leadingAnchor).isActive = true
webView.trailingAnchor.constraint(equalTo: view.trailingAnchor).isActive = true

window.makeKeyAndVisible()
window.rootViewController = vc
```

## Case 1

Default behavior (Before setting the webview's navigation delegate and implementing navigation challenge)

**Results:** 

The web view loads the page, receives a basic auth 401 and shows the response body.

image

## Case 2

After setting the webview's navigation delegate and implementing navigation challenge.

```swift
// code above +
webView.navigationDelegate = self

// and implement relevant method:
extension SceneDelegate: WKNavigationDelegate {
  func webView(_ webView: WKWebView, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
    completionHandler(.performDefaultHandling, nil)
  }
}
```

**Results:** Same as before.

## Case 3

Leaving the implementation blank

```swift
extension SceneDelegate: WKNavigationDelegate {
  func webView(_ webView: WKWebView, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
    // nothing
  }
}
```

**Results:** App crashes with following error:

```
Thread 1: "Completion handler passed to -[sma_shin_ios___debug_dev.SceneDelegate webView:didReceiveAuthenticationChallenge:completionHandler:] was not called"
```

## Case 4

Passing the correct credentials

```swift
extension SceneDelegate: WKNavigationDelegate {
  func webView(_ webView: WKWebView, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
    completionHandler(.performDefaultHandling, nil)
  }
}
```

**Results:** Auth challenge succeeds and http request resolves successfully

image

# TurboNavigator

## Case 1:

You implement the callback and don't do anything:

```swift
func session(
    _ session: Session, didReceiveAuthenticationChallenge challenge: URLAuthenticationChallenge,
    completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
  ) {
      // do nothing..
  }
```

**Result:** App crashes with this error:

```
Thread 1: "Completion handler passed to -[Turbo.ColdBootVisit webView:didReceiveAuthenticationChallenge:completionHandler:] was not called"
```

## Case 2.

You implement the callback and handle the callback with "performDefaultHandling" without providing credentials:

```swift
func session(
    _ session: Session, didReceiveAuthenticationChallenge challenge: URLAuthenticationChallenge,
    completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
  ) {
      completionHandler(.performDefaultHandling, nil)
  }
```

**Result:**  The request is treated as 401 and enters error handling logic

```swift
func session(_ session: Session, didFailRequestForVisitable visitable: Visitable, error: Error) {
  if let turboError = error as? TurboError, case let .http(statusCode) = turboError {
    if statusCode == 401 { // ENTERS THIS CODE PATH
      let authURL = Application.url.appendingPathComponent("/users/sign_in")
      turboNavigator.route(authURL) // infinite loop
    }
  }
}
```

Based on the logic in my app, this just resulted in an infinite loop.

**Note** The exact same thing happens when using

```
completionHandler(.useCredential, nil)
```

## Case 3.

You implement the callback and handle the callback with ".cancelAuthenticationChallenge" without providing credentials:

```swift
func session(
    _ session: Session, didReceiveAuthenticationChallenge challenge: URLAuthenticationChallenge,
    completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
  ) {
      completionHandler(.cancelAuthenticationChallenge, nil)
  }
```

Black screen of death but at least there is no infinite loop 😄 


# Conclusion

When the authentication challenge fails, it gets treated as any other `401` and there is no way to detect that the `401` resulted from a failed authentication challenge once we are in `didFailRequestForVisitable`. This is problematic because if you send the user to `${baseURL}/user/sign_in` which is also under basic auth, then it will result in an infinite loop.

Ideally, `Turbo` would detect that the `www-authenticate: ` is present, and either include that info in `TurboError` so we can treat it differently than other `401`.

here -> https://github.com/hotwired/turbo-ios/blob/main/Source/Visit/ColdBootVisit.swift#L81

But until that is established, it might be better to throw an error to let the user know that we received an http challenge, and that they have to configure the credentials. (Quick feedback)

Something like this:

```swift
func session(
    _ session: Session, didReceiveAuthenticationChallenge challenge: URLAuthenticationChallenge,
    completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
  ) {
    // Uncomment the following lines to use HTTP basic auth
    //
    // completionHandler(
    //     .useCredential,
    //     URLCredential(user: "username", password: "password", persistence: .permanent)
    // )
    fatalError("Your server is behind HTTP basic auth but you did not configure any credentials")
  }
```

In it's current form, this would have to be placed in the user's implementation of the protocol.

**Comment by yanshiyason** [comment]
Here is another proposal: https://github.com/joemasilotti/TurboNavigator/pull/29/files

I am just trying out different approaches. Please let me know your thoughts @joemasilotti 🙇

**Comment by joemasilotti** [comment]
Thank for the detailed write up!

I don't think that Turbo Navigator should make any assumptions about how someone might code their authentication logic. What works for you or me might break something for an existing app relying on basic HTTP auth.

I think the best way to handle this is replicate what turbo-ios is already doing: expose a similar method in the delegate and provide a sensible default implementation. In this scenario, calling the completion block performing default handling.

This means that if someone doesn't implement the `TurboNavigationDelegate` method then the default iOS behavior will occur. And if they implement it then they can do whatever they want.

One thing I'm not sure about yet is if the delegate should expose the web view, Turbo Session, or neither.

```swift
public protocol TurboNavigationDelegate {
    func didReceiveAuthenticationChallenge(_ challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void)
}

public extension TurboNavigationDelegate {
    func didReceiveAuthenticationChallenge(_ challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
        completionHandler(.performDefaultHandling, nil)
    }
}

extension TurboNavigator: SessionDelegate {
    public func session(_ session: Session, didReceiveAuthenticationChallenge challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
        delegate?.didReceiveAuthenticationChallenge(challenge, completionHandler: completionHandler)
    }
}
```

**Comment by yanshiyason** [comment]
@joemasilotti Sorry for the delay,  I updated the PR following the implementation you suggested. I also confirmed that this works in my own iOS app. As a user of the library, my application code looks like this:

```swift
extension Tab: TurboNavigationDelegate {

  // ...

  func didReceiveAuthenticationChallenge(_ challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
    completionHandler(.useCredential, Application.stagingBasicAuthCredentials)
  }
}
```

Please let me know if this PR is good to go.

**Comment by yanshiyason** [comment]
> One thing I'm not sure about yet is if the delegate should expose the web view, Turbo Session, or neither.

I thought I needed access to the web view when I was implementing javascript messaging last week but in the end, I was able to get access to it through these:

```swift 
self.turboNavigator.modalNavigationController.topViewController.visitableView.webView
self.turboNavigator.navigationController.topViewController.visitableView.webView
```

**Review by joemasilotti — APPROVED** [review]

**Pull request merged**